A regression in `deprecate_methods` was introduced in a982a42:

Edouard-chin opened the pull request:
A regression in `deprecate_methods` was introduced in a982a42:

- Refactoring alias_chain to Module#prepend broke the possibility to deprecate class methods since the module
  generated was prepended to the target's instance.
  This patch uses `AS#redefine_method` instead and bind the unbound method to the class
- Fixes #33253
- Credit to @kamipo for the elegant solution

kamipo commented:
How about the simpler below?

```diff
diff --git a/activesupport/lib/active_support/deprecation/method_wrappers.rb b/activesupport/lib/active_support/deprecation/method_wrappers.rb
index 5be893d281..05e15bf6e2 100644
--- a/activesupport/lib/active_support/deprecation/method_wrappers.rb
+++ b/activesupport/lib/active_support/deprecation/method_wrappers.rb
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-require "active_support/core_ext/module/aliasing"
+require "active_support/core_ext/module/redefine_method"
 require "active_support/core_ext/array/extract_options"
 
 module ActiveSupport
@@ -54,23 +54,13 @@ def deprecate_methods(target_module, *method_names)
         deprecator = options.delete(:deprecator) || self
         method_names += options.keys
 
-        mod = Module.new do
-          method_names.each do |method_name|
-            define_method(method_name) do |*args, &block|
-              deprecator.deprecation_warning(method_name, options[method_name])
-              super(*args, &block)
-            end
-
-            case
-            when target_module.protected_method_defined?(method_name)
-              protected method_name
-            when target_module.private_method_defined?(method_name)
-              private method_name
-            end
+        method_names.each do |method_name|
+          method = target_module.instance_method(method_name)
+          target_module.redefine_method(method_name) do |*args, &block|
+            deprecator.deprecation_warning(method_name, options[method_name])
+            method.bind(self).call(*args, &block)
           end
         end
-
-        target_module.prepend(mod)
       end
     end
   end
```

Edouard-chin committed:
A regression in deprecate_methods was introduced in a982a42:

- Refactoring alias_chain to Module#prepend broke the possibility to deprecate class methods since the module
  generated was prepended to the target's instance.

  A suggestion to fix this was  to use `AS#redefine_method` which would solve the
  problem but with the cost of redefining directly the method.
  Decided to go with the same alias_chain implementation as before instead.

- Fixes #33253

Edouard-chin commented:
Thanks @kamipo that's indeed way cleaner. There is one catch though were it won't work

```ruby
module Foo
  def self.bar
  end
end

ActiveSupport::Deprecation.deprecate_methods(Foo, :bar)
```

Since we are using `method = target_module.instance_method(method_name)` but the method is defined on the singleton class

kamipo commented:
Yes, but it is not a regression, originally the `deprecate_methods` won't work against the method defined on the singleton class.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "activesupport", "4.2.10"
end

require "active_support"
require "active_support/deprecation"
require "active_support/testing/deprecation"
require "minitest/autorun"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

class BugTest < Minitest::Test
  include ActiveSupport::Testing::Deprecation

  def test_deprecate_class_method
    mod = Module.new do
      def self.old_method
        "abc"
      end
    end
    ActiveSupport::Deprecation.deprecate_methods(mod, old_method: :new_method)

    warning = /old_method is deprecated and will be removed from Rails \d.\d \(use new_method instead\)/
    assert_deprecated(warning) { assert_equal "abc", mod.old_method }
  end
end
```

```
% ruby 33253.rb
Fetching gem metadata from https://rubygems.org/..............
Resolving dependencies...
Using concurrent-ruby 1.0.5
Using i18n 0.9.5
Using minitest 5.11.3
Using thread_safe 0.3.6
Using tzinfo 1.2.5
Using activesupport 4.2.10
Using bundler 1.16.2
Run options: --seed 52263

# Running:

E

Finished in 0.001092s, 916.0226 runs/s, 0.0000 assertions/s.

  1) Error:
BugTest#test_deprecate_class_method:
NameError: undefined method `old_method' for module `#<Module:0x00007fab46baab78>'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/core_ext/module/aliasing.rb:32:in `alias_method'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/core_ext/module/aliasing.rb:32:in `alias_method_chain'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/deprecation/method_wrappers.rb:34:in `block in deprecate_methods'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/deprecation/method_wrappers.rb:33:in `each'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/deprecation/method_wrappers.rb:33:in `deprecate_methods'
    /Users/kamipo/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activesupport-4.2.10/lib/active_support/deprecation/instance_delegator.rb:19:in `deprecate_methods'
    ac.rb:39:in `test_deprecate_class_method'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

Edouard-chin commented:
Ah ok I didn't know, cool. I have amended my commit with your solution and credited you in the commit. Thanks !

matthewd commented:
On that note, another thought:

```diff
--- a/activesupport/lib/active_support/deprecation/method_wrappers.rb
+++ b/activesupport/lib/active_support/deprecation/method_wrappers.rb
@@ -71,6 +71,7 @@ def deprecate_methods(target_module, *method_names)
         end

         target_module.prepend(mod)
+        target_module.singleton_class.prepend(mod) if target_module.is_a?(target_module)
       end
     end
   end
```

matthewd commented:
Redefining the methods directly in the module has the advantage that it'll work anywhere the module's already been included before the deprecation is added (I haven't verified that, but I think that's the real problem we're seeing here, and may be interesting as a separate test)... but it does also challenge the idea that `alias_method_chain` is replaceable with `prepend` -- as well as doing not-great things to performance.

Hrmm. :confused:

kamipo commented:
We can't replace `alias_method_chain` with `prepend` in this case as long as keeping the compatibility with 4.2.
The `prepend` version won't pass the following test case (4.2 and the `redefine_method` version would pass though).

```ruby
  def test_deprecate_class_method2                                                                                                          
    mod = Module.new do                                                                                                                     
      def old_method                                                                                                                        
        "abc"                                                                                                                               
      end                                                                                                                                   
    end                                                                                                                                     
    @klass.extend mod                                                                                                                       
    ActiveSupport::Deprecation.deprecate_methods(mod, old_method: :new_method)                                                              
                                                                                                                                            
    warning = /old_method is deprecated and will be removed from Rails \d.\d \(use new_method instead\)/                                    
    assert_deprecated(warning) { assert_equal "abc", @klass.old_method }                                                                    
  end                                                                                                                                       
```

rafaelfranca commented:
would using the `alias_method_chain` implementation be better in this case? It would not require us to redefine the method.

Edouard-chin commented:
If we want to keep compatibility with 4.2 then I suppose this is our best option. I'll push a commit soon

kamipo commented:
Should we test accessibility to `"#{aliased_method}_with_deprecation` and `"#{aliased_method}_without_deprecation`?
It is the difference between `redefine_method` and `alias_method_chain`.

Edouard-chin commented:
Thanks for the quick review.

I'm not sure if I understand your comment @kamipo. Do you want to ensure that `with_..` and `without_..` methods have the same accessibility as the original one?

Isn't the 2 test cases here enough https://github.com/rails/rails/blob/6db5b6dc4322d043db4d6eff2cc49de86baf1595/activesupport/test/deprecation/method_wrappers_test.rb#L47-L57 ?

kamipo commented:
I meant the difference is here:

```diff
diff --git a/activesupport/test/deprecation/method_wrappers_test.rb b/activesupport/test/deprecation/method_wrappers_test.rb
index e9d5295945..b04bce7a11 100644
--- a/activesupport/test/deprecation/method_wrappers_test.rb
+++ b/activesupport/test/deprecation/method_wrappers_test.rb
@@ -82,4 +82,12 @@ def old_method
     warning = /old_method is deprecated and will be removed from Rails \d.\d \(use new_method instead\)/
     assert_deprecated(warning) { assert_equal "abc", @klass.old_method }
   end
+
+  def test_method_with_without_deprecation_is_exposed
+    ActiveSupport::Deprecation.deprecate_methods(@klass, old_method: :new_method)
+
+    warning = /old_method is deprecated and will be removed from Rails \d.\d \(use new_method instead\)/
+    assert_deprecated(warning) { assert_equal "abc", @klass.new.old_method_with_deprecation }
+    assert_equal "abc", @klass.new.old_method_without_deprecation
+  end
 end
```

This won't be passed with `redefine_method`, but would be pass with `alias_method_chain`.

The reason for using `redefine_method` I proposed is that I regarded the internal methods exposed as a limitation rather than a specification.
If it is a limitation due to older implementation, we could just avoid it now.

Edouard-chin commented:
Ahh ok , yes of course that test makes complete sense. Let me apply the patch. (BTW thanks a lot for taking the time to write a diff 😍)

kamipo commented on activesupport/test/deprecation/method_wrappers_test.rb:
```diff
@@ -55,4 +55,39 @@ def test_deprecate_methods_private_method
+  end
+
+   def test_method_with_without_deprecation_is_exposed
+    ActiveSupport::Deprecation.deprecate_methods(@klass, old_method: :new_method)
```
Need indented one more space here.
